tracer/src/Datadog.Trace/DatabaseMonitoring/DatabaseMonitoringPropagator.cs

```diff
@@ -35,76 +35,129 @@ internal static class DatabaseMonitoringPropagator
-                {
-                    propagatorStringBuilder.Append(',').Append(SqlCommentVersion).Append("='").Append(Uri.EscapeDataString(versionTag)).Append('\'');
-                }
+            if (span.Context.TraceContext?.ServiceVersion is { } versionTag)
+            {
+                propagatorStringBuilder.Append(',').Append(SqlCommentVersion).Append("='").Append(Uri.EscapeDataString(versionTag)).Append('\'');
+            }
+
+            var traceParentInjected = false;
+            // For SqlServer & Oracle we don't inject the traceparent to avoid affecting performance, since those DBs generate a new plan for any query changes
+            if (propagationLevel == DbmPropagationLevel.Full
+             && integrationId is not (IntegrationId.SqlClient or IntegrationId.Oracle))
+            {
+                traceParentInjected = true;
+                propagatorStringBuilder.Append(',').Append(W3CTraceContextPropagator.TraceParentHeaderName).Append("='").Append(W3CTraceContextPropagator.CreateTraceParentHeader(span.Context)).Append('\'');
+            }
 
-                // For SqlServer & Oracle we don't inject the traceparent to avoid affecting performance, since those DBs generate a new plan for any query changes
-                if (propagationStyle == DbmPropagationLevel.Full
-                 && integrationId is not (IntegrationId.SqlClient or IntegrationId.Oracle))
+            propagatorStringBuilder.Append("*/");
+
+            // modify the command to add the comment
+            var commandText = command.CommandText ?? string.Empty;
+            var propagationComment = StringBuilderCache.GetStringAndRelease(propagatorStringBuilder);
+            if (ShouldAppend(integrationId, commandText))
+            {
+                command.CommandText = $"{commandText} {propagationComment}";
+            }
+            else
+            {
+                // prepending the propagation comment is the preferred way,
+                // as this protects it from being truncated by the character limit if the command is very long.
+                command.CommandText = $"{propagationComment} {commandText}";
+            }
+
+            return traceParentInjected;
+        }
+
+        internal static bool ShouldAppend(IntegrationId integrationId, string commandText)
+        {
+            // pg_hint_plan allows setting hints for the execution plan as the *first* comment. If such a hint is present,
+            // we need to append our comment rather than prepend, to avoid invalidating the hint
+            // see https://pg-hint-plan.readthedocs.io/en/latest/hint_details.html#syntax-and-placement
+            return integrationId == IntegrationId.Npgsql && StartsWithHint(commandText);
+        }
+
+        /// <summary>
+        /// Detect if a command contains a pg_hint_plan.
+        /// Basically equivalent to `commandText.TrimStart().StartsWith("/*+")` except it avoids the new string allocation that Trim does.
```

Comment:
Since we can now use `Span<char>`, you should be able to implement this entire method without the extra string allocation with simply
```csharp
return commandText.AsSpan().TrimStart().StartsWith("/*+");
```

Reply:
Very nice